### PR TITLE
Add return type to Canonical_Formatter::format() (to stop promoter breaking Woocommerce Checkout)

### PR DESCRIPTION
**Pull request by justlevine**
Tl;dr adds return type of `string` to LineFormatter::format, so it matches what monolog expects.

`Declaration of Tribe\Log\Canonical_Formatter::format(array $record) must be compatible with Monolog\Formatter\LineFormatter::format(array $record): string`.

---
Longer version:

The newer versions of Event Tickets Plus was breaking Woocommerce, because Promotor was throwing an exception. `['ticket_purchased', 'License Key must be present].

Normally this should be ignored (I for example use the nonprofit license, so we dont have promoter), but issue with monolog were causing a fatal error:

`PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Monolog\Handler\ErrorLogHandler::__construct() must be of the type int, null given, called in /var/www/thegsc/wp-content/plugins/the-events-calendar/common/src/Tribe/Log/Service_Provider.php on line 73.`

That's a separate issue, and i'm not familiar enough with Tribe's Logger, but I changed ` new ErrorLogHandler( null, $level_threshold );` to `new ErrorLongHandler( 0, $level_threshold)`, since [thats the default value provided by Monolog](https://github.com/Seldaek/monolog/blob/bb99e4c699bd5abe2609cace3c2daf0e45fdb673/src/Monolog/Handler/ErrorLogHandler.php#L25)...

Then I got hit with another error, and that's what this fix addresses

```
PHP Fatal error:  Declaration of Tribe\Log\Canonical_Formatter::format(array $record) must be compatible with Monolog\Formatter\LineFormatter::format(array $record): string in /var/www/thegsc/wp-content/plugins/the-events-calendar/common/src/Tribe/Log/Canonical_Formatter.php on line 27
[08-Oct-2020 16:30:43 UTC] PHP Stack trace:
[08-Oct-2020 16:30:43 UTC] PHP   1. {main}() /var/www/thegsc/index.php:0
[08-Oct-2020 16:30:43 UTC] PHP   2. require() /var/www/thegsc/index.php:17
[08-Oct-2020 16:30:43 UTC] PHP   3. require_once() /var/www/thegsc/wp-blog-header.php:19
[08-Oct-2020 16:30:43 UTC] PHP   4. do_action() /var/www/thegsc/wp-includes/template-loader.php:13
[08-Oct-2020 16:30:43 UTC] PHP   5. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP   6. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP   7. WC_AJAX::do_wc_ajax() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP   8. do_action() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-ajax.php:89
[08-Oct-2020 16:30:43 UTC] PHP   9. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  10. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  11. WC_AJAX::checkout() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP  12. WC_Checkout->process_checkout() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-ajax.php:464
[08-Oct-2020 16:30:43 UTC] PHP  13. WC_Checkout->process_order_payment() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-checkout.php:1159
[08-Oct-2020 16:30:43 UTC] PHP  14. WC_Gateway_COD->process_payment() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-checkout.php:956
[08-Oct-2020 16:30:43 UTC] PHP  15. Automattic\WooCommerce\Admin\Overrides\Order->update_status() /var/www/thegsc/wp-content/plugins/woocommerce/includes/gateways/cod/class-wc-gateway-cod.php:326
[08-Oct-2020 16:30:43 UTC] PHP  16. Automattic\WooCommerce\Admin\Overrides\Order->save() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-order.php:333
[08-Oct-2020 16:30:43 UTC] PHP  17. Automattic\WooCommerce\Admin\Overrides\Order->status_transition() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-order.php:221
[08-Oct-2020 16:30:43 UTC] PHP  18. do_action() /var/www/thegsc/wp-content/plugins/woocommerce/includes/class-wc-order.php:373
[08-Oct-2020 16:30:43 UTC] PHP  19. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  20. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  21. Tribe__Tickets_Plus__Commerce__WooCommerce__Main->delayed_ticket_generation() /var/www/thegsc/wp-includes/class-wp-hook.php:289
[08-Oct-2020 16:30:43 UTC] PHP  22. Tribe__Tickets_Plus__Commerce__WooCommerce__Main->generate_tickets() /var/www/thegsc/wp-content/plugins/event-tickets-plus/src/Tribe/Commerce/WooCommerce/Main.php:795
[08-Oct-2020 16:30:43 UTC] PHP  23. do_action() /var/www/thegsc/wp-content/plugins/event-tickets-plus/src/Tribe/Commerce/WooCommerce/Main.php:973
[08-Oct-2020 16:30:43 UTC] PHP  24. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  25. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  26. Tribe__Tickets_Plus__Commerce__WooCommerce__Promoter_Observer->ticket_created() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP  27. Tribe__Tickets_Plus__Commerce__WooCommerce__Promoter_Observer->trigger() /var/www/thegsc/wp-content/plugins/event-tickets-plus/src/Tribe/Commerce/WooCommerce/Promoter_Observer.php:70
[08-Oct-2020 16:30:43 UTC] PHP  28. do_action() /var/www/thegsc/wp-content/plugins/event-tickets-plus/src/Tribe/Commerce/WooCommerce/Promoter_Observer.php:95
[08-Oct-2020 16:30:43 UTC] PHP  29. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  30. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  31. Tribe\Tickets\Promoter\Triggers\Factory->build_attendee() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP  32. do_action() /var/www/thegsc/wp-content/plugins/event-tickets/src/Tribe/Promoter/Triggers/Factory.php:40
[08-Oct-2020 16:30:43 UTC] PHP  33. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  34. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  35. Tribe\Tickets\Promoter\Triggers\Dispatcher->trigger() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP  36. do_action() /var/www/thegsc/wp-content/plugins/event-tickets/src/Tribe/Promoter/Triggers/Dispatcher.php:90
[08-Oct-2020 16:30:43 UTC] PHP  37. WP_Hook->do_action() /var/www/thegsc/wp-includes/plugin.php:478
[08-Oct-2020 16:30:43 UTC] PHP  38. WP_Hook->apply_filters() /var/www/thegsc/wp-includes/class-wp-hook.php:311
[08-Oct-2020 16:30:43 UTC] PHP  39. Tribe\Log\Service_Provider->dispatch_log() /var/www/thegsc/wp-includes/class-wp-hook.php:287
[08-Oct-2020 16:30:43 UTC] PHP  40. Tribe__Container->make() /var/www/thegsc/wp-content/plugins/the-events-calendar/common/src/Tribe/Log/Service_Provider.php:130
[08-Oct-2020 16:30:43 UTC] PHP  41. Tribe\Log\Service_Provider->build_logger() /var/www/thegsc/wp-content/plugins/the-events-calendar/common/vendor/lucatume/di52/src/tad/DI52/Container.php:279
[08-Oct-2020 16:30:43 UTC] PHP  42. spl_autoload_call() /var/www/thegsc/wp-content/plugins/the-events-calendar/common/src/Tribe/Log/Service_Provider.php:85
[08-Oct-2020 16:30:43 UTC] PHP  43. Composer\Autoload\ClassLoader->loadClass() /var/www/thegsc/wp-content/plugins/the-events-calendar/common/src/Tribe/Log/Service_Provider.php:85
[08-Oct-2020 16:30:43 UTC] PHP  44. Composer\Autoload\includeFile() /var/www/thegsc/wp-content/plugins/event-tickets/vendor/composer/ClassLoader.php:322
[08-Oct-2020 16:30:43 UTC] PHP  45. include() /var/www/thegsc/wp-content/plugins/event-tickets/vendor/composer/ClassLoader.php:444
```

### CODE REVIEW

**Comment by sc0ttkclark** [comment]
Thanks for the PR here! We'll be following up on this during our release planning process in the upcoming week or so.

**Comment by justlevine** [comment]
Thanks @sc0ttkclark !
In case it wasnt clear from my long post, this alone doesnt fix the fatal error during wc checkout. That's still being caused by `Tribe\Log\Service_Provider::build_logger():73`, since the first parameter of `ErrorLogHandler()` is set to `null`, and not an int. While, i fixed it locally by switching `null` with `0`, I'm assuming the intent was to inherit `self::OPERATING_SYSTEM` from [Monolog](https://github.com/Seldaek/monolog/blob/bb99e4c699bd5abe2609cace3c2daf0e45fdb673/src/Monolog/Handler/ErrorLogHandler.php#L25), which I'm not sure the best approach on how to fix (passing 'null' isnt the same as leaving the parameter undefined)...

**Comment by sc0ttkclark** [comment]
@justlevine Thanks for the further clarification, our team is actually looking at this now to see what our plan is to address it. Thanks for your patience as we review this.

**Comment by bordoni** [comment]
@justlevine do you have any other version of Monolog being used on your install? Maybe some other plugin is including it too. Because I am looking at the method: https://github.com/Seldaek/monolog/blob/1.x/src/Monolog/Formatter/LineFormatter.php#L68-L113

And it seems like it follows the same structure as the one we have on our code. We are using version `1.X` if any other Monolog is loaded using version `2.X` it would cause that error.

**Comment by justlevine** [comment]
You're correct 😳. 
When disabling plugins, I forgot there were mu-plugins on this install... one of them is using Monolog v2.x. The fatal error is only happening now because my TEC instance is finally logging an error (from the missing Promoter license).

In lieu of prefixing your composer dependencies to prevent conflicts, perhaps you can tweak how TEC checks for Promoter, eg. returning instead of throwing/logging an Exception.

Also, a quick check of my disabled plugins shows Woocommerce Bookings also uses v2.x (although the majority of other plugins use v1.x, so upgrading the dependency would probably cause more harm than good at this point). Unless the Promoter exception is changed, I'd suggest adding WC Bookings to your list of conflicting plugins.

Either way, I guess you can disregard this PR (ps: the issue with passing null to `ErrorLogHandler()` [still exists in version 1.x](https://github.com/Seldaek/monolog/blob/1.x/src/Monolog/Handler/ErrorLogHandler.php#L36) ). 

Thanks for the quick response, and apologies for wasting your time!

**Comment by borkweb** [comment]
@justlevine - just following up on this ticket to wrap things up. We've found a solution that changes our Monolog integration to be compatible with both 1.x _and_ 2.x which should work around the issue that you reported! We are aiming to release the fix in the next version of TEC (likely next week).

Thank you again for spotting this issue and bouncing ideas around!
